singleton/controller.go

```diff
@@ -36,3 +38,33 @@ func Source() source.Source {
 		},
 	})
 }
+
+func ChannelSource[T client.Object](ctx context.Context, objectChan <-chan watch.Event) source.Source {
+	eventSource := make(chan event.GenericEvent, 1)
+
+	go func(ctx context.Context, objectChan <-chan watch.Event, eventSource chan event.GenericEvent) {
+
+		for {
+			select {
+			case delta, ok := <-objectChan:
+				if !ok {
+					return
+				}
+				// Convert the delta to GenericEvent
+				eventSource <- event.GenericEvent{
+					Object: delta.Object.(T),
+				}
+			case <-ctx.Done():
+				return
+			}
+		}
+	}(ctx, objectChan, eventSource)
+
+	return source.Channel(eventSource, handler.Funcs{
```

Comment:
Why can't we read directly from the watch channel here?

Reply:
We are now using the watch channel directly